File: DraftRight/Utils/Logger.swift
import Foundation

/// Centralized file logger for DraftRight.
/// Logs to ~/Library/Logs/DraftRight/draftright.log with timestamps and categories.
enum DRLogger {
    /// Toggle logging on/off. Persisted in UserDefaults.
    static var isEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: "draftright.loggingEnabled") }
        set { UserDefaults.standard.set(newValue, forKey: "draftright.loggingEnabled") }
    }

    private static let logDir: URL = {
        let dir = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Logs/DraftRight")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }()

    private static let logFile: URL = logDir.appendingPathComponent("draftright.log")

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        return f
    }()

    enum Category: String {
        case app = "APP"
        case auth = "AUTH"
        case ax = "AX"
        case monitor = "MONITOR"
        case panel = "PANEL"
        case api = "API"
        case settings = "SETTINGS"
    }

    /// Log severity. Lines are tagged `[LEVEL]` so failures are skimmable and
    /// greppable (`grep "\[ERROR\]"`) instead of buried in same-level category
    /// chatter. `.warn` = something degraded but handled; `.error` = an
    /// operation failed.
    enum Level: String {
        case info = "INFO"
        case warn = "WARN"
        case error = "ERROR"
    }

    static func log(_ message: String, category: Category = .app, level: Level = .info) {
        // WARN/ERROR are always recorded, even when the user has logging
        // toggled off — a bug report shouldn't be blank for the lines that
        // matter most. Only routine INFO chatter honors the off-switch.
        guard isEnabled || level != .info else { return }

        let timestamp = dateFormatter.string(from: Date())
        // Pad the level so the [CATEGORY] column stays aligned across lines.
        let levelTag = level.rawValue.padding(toLength: 5, withPad: " ", startingAt: 0)
        let line = "[\(timestamp)] [\(levelTag)] [\(category.rawValue)] \(message)\n"

        if let data = line.data(using: .utf8) {
            if let handle = try? FileHandle(forWritingTo: logFile) {
                handle.seekToEndOfFile()
                handle.write(data)
                handle.closeFile()
            } else {
                try? data.write(to: logFile)
            }
        }

        #if DEBUG
        print(line, terminator: "")
        #endif
    }

    /// Log a handled-but-degraded condition at `.warn`.
    static func warn(_ message: String, category: Category = .app) {
        log(message, category: category, level: .warn)
    }

    /// Log a failed operation at `.error`.
    static func error(_ message: String, category: Category = .app) {
        log(message, category: category, level: .error)
    }

    /// Returns the log file path for display in Settings
    static var logFilePath: String { logFile.path }

    /// Returns the last N lines of the log
    static func tail(lines: Int = 100) -> String {
        guard let content = try? String(contentsOf: logFile, encoding: .utf8) else { return "(no logs)" }
        let allLines = content.components(separatedBy: "\n")
        let lastLines = allLines.suffix(lines)
        return lastLines.joined(separator: "\n")
    }

    /// Clears the log file
    static func clear() {
        try? "".write(to: logFile, atomically: true, encoding: .utf8)
    }
}

File: DraftRightWindows/DraftRightWindows/Services/Logger.cs
using System;
using System.IO;

namespace DraftRightWindows.Services;

public static class DRLogger
{
    private static readonly string LogDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "DraftRight", "Logs");

    private static readonly string LogFile = Path.Combine(LogDir, "draftright.log");
    private static readonly object Lock = new();

    public enum Category { APP, AUTH, API, PANEL, SETTINGS, HOTKEY }

    /// <summary>
    /// Log severity. Lines are tagged <c>[LEVEL]</c> so failures are skimmable
    /// and greppable (<c>Select-String "\[ERROR\]"</c>) instead of buried in
    /// same-level category chatter. <see cref="Level.WARN"/> = something
    /// degraded but handled; <see cref="Level.ERROR"/> = an operation failed.
    /// </summary>
    public enum Level { INFO, WARN, ERROR, OFF }

    /// <summary>
    /// Master switch — when false, Log() short-circuits without writing.
    /// Settings.LoggingEnabled is mirrored here at startup and on toggle so
    /// hot paths don't have to reach into SettingsService.
    /// </summary>
    public static bool IsEnabled { get; set; } = true;

    /// <summary>
    /// Minimum severity to write, set remotely by the admin portal and pushed
    /// to clients via <c>/health</c>'s <c>client_log_level</c>. The master
    /// verbosity control: <see cref="Level.INFO"/> logs everything (default),
    /// and <see cref="Level.OFF"/> is the absolute kill-switch (silences even
    /// errors — for privacy/compliance). <see cref="Level.OFF"/> is only ever
    /// used as this threshold, never passed as a line's level.
    /// </summary>
    public static Level MinLevel { get; set; } = Level.INFO;

    static DRLogger()
    {
        Directory.CreateDirectory(LogDir);
    }

    public static void Log(string message, Category category = Category.APP, Level level = Level.INFO)
    {
        // Remote admin threshold (master): drop anything below it; OFF drops all.
        if (level < MinLevel) return;
        // WARN/ERROR are always recorded, even when the user has logging
        // toggled off — a bug report shouldn't be blank for the lines that
        // matter most. Only routine INFO chatter honors the off-switch.
        if (!IsEnabled && level == Level.INFO) return;

        // Pad the level so the [CATEGORY] column stays aligned across lines.
        var line = $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff}] [{level,-5}] [{category}] {message}";
        lock (Lock)
        {
            File.AppendAllText(LogFile, line + Environment.NewLine);
        }
#if DEBUG
        System.Diagnostics.Debug.WriteLine(line);
#endif
    }

    /// <summary>Log a handled-but-degraded condition at <see cref="Level.WARN"/>.</summary>
    public static void Warn(string message, Category category = Category.APP) =>
        Log(message, category, Level.WARN);

    /// <summary>Log a failed operation at <see cref="Level.ERROR"/>.</summary>
    public static void Error(string message, Category category = Category.APP) =>
        Log(message, category, Level.ERROR);

    /// <summary>
    /// Applies the backend's <c>client_log_level</c> ("off" | "errors" |
    /// "warnings" | "info") as the <see cref="MinLevel"/> threshold. Unknown or
    /// empty values fall back to full logging. No-op (and silent) when the
    /// level is unchanged, so the ~30s health poll doesn't spam the log; only a
    /// genuine change is announced. Safe to call from any thread.
    /// </summary>
    public static void SetMinLevelFromServer(string? value)
    {
        var newLevel = (value?.Trim().ToLowerInvariant()) switch
        {
            "off" => Level.OFF,
            "errors" or "error" => Level.ERROR,
            "warnings" or "warning" or "warn" => Level.WARN,
            "info" or "" or null => Level.INFO,
            _ => Level.INFO,
        };
        if (newLevel == MinLevel) return;

        var old = MinLevel;
        // Announce the change so on/off transitions are traceable. When
        // narrowing (e.g. → OFF) record it under the OLD threshold first so the
        // announcement itself isn't dropped; when widening, set then record.
        if (newLevel > old)
        {
            Log($"Client log level changed: {old} -> {newLevel} (server '{value}')", Category.APP, Level.WARN);
            MinLevel = newLevel;
        }
        else
        {
            MinLevel = newLevel;
            Log($"Client log level changed: {old} -> {newLevel} (server '{value}')", Category.APP, Level.WARN);
        }
    }

    public static string LogFilePath => LogFile;
}
